service/internal/asg/asg.go

```diff
@@ -0,0 +1,298 @@
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+
+	Stack        string
+	TagKey       string
+	TagValueFunc func(cr key.LabelsGetter) string
+}
+
+type ASG struct {
+	k8sClient k8sclient.Interface
+
+	asgsCache      *cache.ASGs
+	instancesCache *cache.Instances
+
+	stack        string
+	tagKey       string
+	tagValueFunc func(cr key.LabelsGetter) string
+}
+
+func New(config Config) (*ASG, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	if config.Stack == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Stack must not be empty", config)
+	}
+	if config.TagKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagKey must not be empty", config)
+	}
+	if config.TagValueFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagValueFunc must not be empty", config)
+	}
+
+	a := &ASG{
+		k8sClient: config.K8sClient,
+
+		asgsCache:      cache.NewASGs(),
+		instancesCache: cache.NewInstances(),
+
+		stack:        config.Stack,
+		tagKey:       config.TagKey,
+		tagValueFunc: config.TagValueFunc,
+	}
+
+	return a, nil
+}
+
+func (a *ASG) Drainable(ctx context.Context, obj interface{}) (string, error) {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	// In order to fetch all the relevant ASGs we have to fetch all relevant
+	// instances. The Autoscaling API does not provide tag filters but only
+	// filtering by ASG names. Therefore we get the ASG names from the instances,
+	// because the EC2 API has tag filter support.
+	instances, err := a.cachedInstances(ctx, cr)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	// Having the EC2 instances we can parse the ASG names from the instance tags
+	// and fetch the ASGs in order to get the lifecycle hook information we are
+	// interested it.
+	var asgs []*autoscaling.Group
+	{
+		var names []string
+		{
+			m := map[string]struct{}{}
+
+			for _, i := range instances {
+				m[asgNameFromInstance(i)] = struct{}{}
+			}
+
+			for k := range m {
+				names = append(names, k)
+			}
+		}
+
+		asgs, err = a.cachedASGs(ctx, cr, names)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+	}
+
+	// The user asks for the next drainable ASG. There is 1 ASG only, for
+	// instance for Node Pools and Single Master setups. There are 3 ASGs in HA
+	// Masters setups. In case there are multiple ASGs we want to be careful and
+	// only drain instances of one of the ASGs at a time. Further, we only want
+	// to drain the next ASG, in case each of the other ASGs contains at least
+	// one healthy instance being InService. This aims to ensure that an HA
+	// Masters setup is most reliable.
+	if len(asgs) > 1 {
+		var c int
+
+		for _, a := range asgs {
+			for _, i := range a.Instances {
+				if *i.LifecycleState == autoscaling.LifecycleStateInService {
+					c++
+					break
+				}
+			}
+		}
```

Comment:
this works weirdly in my head, I guess we need to test if it behaves how we want, maybe it's correct it's just my head that does not compute :rofl:

Reply:
OK I wrote it down and it looks fine :)